src/network/matrix/matrix.ts

```diff
@@ -0,0 +1,175 @@
+/// <reference path="./matrix.type.d.ts" />
+import sdk, { MatrixClient } from 'matrix-js-sdk'
+import mitt from 'mitt'
+export const endpoint = 'https://matrix.vampire.rip'
+type MatrixClient = typeof MatrixClient extends { new (...args: any[]): infer U } ? U : never
+function any(x: any): any {
+    return x
+}
+type MatrixLoginCred = Record<'user_id' | 'device_id' | 'access_token' | 'home_server', string>
+/**
+ * Login a Matrix Account
+ * @param username
+ * @param password
+ */
+export async function loginMatrixAccount(
+    client: MatrixClient,
+    username: string,
+    password: string,
+): Promise<[MatrixLoginCred, MatrixClient]> {
+    const cred: MatrixLoginCred = await any(client).loginWithPassword(username, password)
+    client.startClient({})
+    return [cred, client]
+}
+/**
+ * Register a Matrix Account
+ */
+export function registerMatrixAccount(
+    client: MatrixClient,
+    username: string,
+    password: string,
+): Promise<[MatrixLoginCred, MatrixClient]> {
+    const u = undefined
+    return new Promise<[MatrixLoginCred, MatrixClient]>(async (resolve, reject) => {
+        try {
+            any(client).register(username, password, u, u, u, u, false, (x: any) => {
+                const session = x.data.session
+                if (!session) return reject(x)
+                any(client).register(username, password, session, { type: 'm.login.dummy' }, u, u, false, () =>
+                    loginMatrixAccount(client, username, password).then(resolve, reject),
+                )
+            })
+        } catch (e) {
+            reject(e)
+        }
+    })
+}
+export function createMatrixRoom(client: MatrixClient, visibility: 'public' | 'private', room_alias_name: string) {
+    return any(client).createRoom({ room_alias_name, visibility })
+}
+interface MatrixMessageTypes {
+    'maskbook.hello.world': { message: string }
+    'maskbook.hello.world2': number
+}
+interface MatrixMessageEvent<T extends keyof MatrixMessageTypes> {
+    meta: {
+        matrix: {
+            sender: string
+            origin_server_ts: number
+            event_id: string
+            room_id: string
+            room_name?: string
+        }
+    }
+    type: T
+    payload: MatrixMessageTypes[T]
+}
+type MatrixMessageListener<T extends keyof MatrixMessageTypes> = (event: MatrixMessageEvent<T>) => void
+interface PatchedEvent extends Omit<MatrixEvent, 'event'> {
+    event: {
+        type: string
+        sender: string
+        content: object
+        origin_server_ts: number
+        unsigned: unknown & {
+            age: number
+        }
+        event_id: string
+        room_id: string
+    }
+}
+export class MatrixMessage {
+    public userID: null | string = null
+    constructor(public matrixClient: MatrixClient) {
```

Comment:
Async logic in the `constructor` is not good.

Reply:
If the client is not started the state might be mrssed